rescan.go

```diff
@@ -573,25 +538,51 @@ func rescan(chain ChainSource, options ...RescanOption) error {
 		// If the filter is nil, then this either means that we don't
 		// have any peers to fetch this filter from, or the peer(s) that
 		// we're trying to fetch from are in the progress of a re-org.
+		// In any case, we'll attempt to retry fetching the filter until
+		// success or we determine it belongs to a stale block.
 		if blockFilter == nil {
-			// TODO(halseth): this is racy, as blocks can come in
-			// before we refetch.
-			resetBlockReFetchTimer(header, uint32(curStamp.Height))
+			// Only add it to the queue if it hasn't been added yet.
+			if !isRetry {
+				retryBlocks = append(retryBlocks, ntfn)
+			}
+
+			log.Debugf("Retrying filter fetch for height=%v, "+
+				"hash=%v after %v", ntfn.Height(),
+				newHeader.BlockHash(), blockRetryInterval)
+
+			blockRetrySignal = time.After(blockRetryInterval)
+
 			return nil
 		}
 
+		// Otherwise, we've obtained the filter, so we'll proceed to
+		// retrieve any matches and notify them.
 		err = notifyBlockWithFilter(
-			chain, ro, &curHeader, &curStamp, blockFilter,
+			chain, ro, &newHeader, &newStamp, blockFilter,
 		)
 		if err != nil {
 			return err
 		}
 
-		// We've successfully fetched this current block, so we'll reset
-		// the retry timer back to nil.
-		if blockReFetchTimer != nil {
-			blockReFetchTimer.Stop()
-			blockReFetchTimer = nil
+		// With the notifications sent, we can update our in-memory
+		// state.
+		curHeader = newHeader
+		curStamp = newStamp
+
+		// If this was a retry, we'll proceed to the next if any.
+		if isRetry {
+			retryBlocks[0] = nil
+			retryBlocks = retryBlocks[1:]
+			if len(retryBlocks) > 0 {
+				next := retryBlocks[0]
+				nextHeader := next.Header()
+
+				log.Debugf("Retrying filter fetch for height=%v, "+
+					"hash=%v after %v", next.Height(),
+					nextHeader.BlockHash(), blockRetryInterval)
+
+				blockRetrySignal = time.After(blockRetryInterval)
```

Comment:
If we are successful, is a new retry interval needed or can it continue directly with the rest of the list?

Reply:
We could proceed without another signal, but the reasoning behind it is to allow the possibility of a disconnected notification to be sent to the rescan so that it can stop retrying.